sdks/python/apache_beam/io/mongodbio.py

```diff
@@ -194,18 +212,110 @@ def display_data(self):
     res['mongo_client_spec'] = self.spec
     return res
 
-  def _get_avg_document_size(self):
+  def _get_split_keys(self, desired_chunk_size_in_mb, start_pos, end_pos):
+    # if desired chunk size smaller than 1mb, use mongodb default split size of
+    # 1mb
+    if desired_chunk_size_in_mb < 1:
+      desired_chunk_size_in_mb = 1
+    if start_pos >= end_pos:
+      # single document not splittable
+      return []
     with MongoClient(self.uri, **self.spec) as client:
-      size = client[self.db].command('collstats', self.coll).get('avgObjSize')
-      if size is None or size <= 0:
-        raise ValueError(
-            'Collection %s not found or average doc size is '
-            'incorrect', self.coll)
-      return size
-
-  def _get_document_count(self):
+      name_space = '%s.%s' % (self.db, self.coll)
+      return (client[self.db].command(
+          'splitVector',
+          name_space,
+          keyPattern={'_id': 1},
+          min={'_id': start_pos},
+          max={'_id': end_pos},
+          maxChunkSize=desired_chunk_size_in_mb)['splitKeys'])
+
+  def _merge_id_filter(self, range_tracker):
+    all_filters = self.filter.copy()
+    if '_id' in all_filters:
+      id_filter = all_filters['_id']
+      id_filter['$gte'] = (
+          max(id_filter['$gte'], range_tracker.start_position())
+          if '$gte' in id_filter else range_tracker.start_position())
+
+      id_filter['$lt'] = (min(id_filter['$lt'], range_tracker.stop_position())
+                          if '$lt' in id_filter else
+                          range_tracker.stop_position())
+    else:
+      all_filters.update({
+          '_id': {
+              '$gte': range_tracker.start_position(),
+              '$lt': range_tracker.stop_position()
+          }
+      })
+    return all_filters
+
+  def _get_head_document_id(self, sort_order):
     with MongoClient(self.uri, **self.spec) as client:
-      return max(client[self.db][self.coll].count_documents(self.filter), 0)
+      cursor = client[self.db][self.coll].find(filter={}, projection=[]).sort([
+          ('_id', sort_order)
+      ]).limit(1)
+      try:
+        return cursor[0]['_id']
+      except IndexError:
+        raise ValueError('Empty Mongodb collection')
+
+
+class _ObjectIdHelper(object):
+  """A Utility class to bson object ids."""
+
+  @classmethod
+  def id_to_int(cls, id):
+    # converts object id binary to integer
```

Comment:
Please add a proper doc comment including variables if possible. (even though containing class is private, better to properly describe these functions that contain non-trivial byte manipulations).